File: src/game/quests/quest-engine.ts
// === Quest Engine — objective evaluation + reward application ===
//
// Layer 3 (Event System). Ticked each game tick alongside LogicNodeEngine.
// Reads events from the bus, checks active quest objectives, emits
// QuestComplete / QuestUpdate / QuestFail events.
//
// Adding a new objective type = one entry in OBJECTIVE_CHECKERS.
// Adding a new reward type = one entry in REWARD_APPLIERS.

import {EventTag} from '../event-types';
import type {EventBus} from '../event-bus';
import {
	type PlayerState, type GameState, type AnySettlement,
} from '../state';
import {torusDist} from '../torus';
import {
	removeItem, addItem, makeItem,
} from '../items';
import type {Quest, QuestObjective, QuestReward} from './quest-types';

// ── Objective checker registry (data-driven) ──

type ObjectiveChecker = (
	object: QuestObjective,
	bus: EventBus,
	player: PlayerState,
	state: GameState,
	mapWidth: number,
	mapHeight: number,
) => boolean;

const OBJECTIVE_CHECKERS: Record<string, ObjectiveChecker> = {
	visit_cell(object, _bus, player, _state, mapWidth, mapHeight) {
		if (object.type !== 'visit_cell') {
			return false;
		}

		const d = torusDist(player.x, player.y, object.x, object.y, mapWidth, mapHeight);
		return d <= object.radius;
	},

	find_location(object, bus) {
		if (object.type !== 'find_location') {
			return false;
		}

		// Checked via a dedicated SubworldProximity event (future)
		// For now, check if player entered the right cell
		for (const ev of bus.lastTickEvents) {
			if (ev.tag === EventTag.PlayerMove
				&& 'toX' in ev && 'toY' in ev
				&& ev.toX === object.cellX && ev.toY === object.cellY) {
				return true;
			}
		}

		return false;
	},

	deliver_items(object, _bus, player, state, mapWidth, mapHeight) {
		if (object.type !== 'deliver_items') {
			return false;
		}

		// Player must be at target settlement AND have items
		const target = findSettlement(state, object.targetSettlementId);
		if (!target) {
			return false;
		}

		const d = torusDist(player.x, player.y, target.x, target.y, mapWidth, mapHeight);
		if (d > 3) {
			return false;
		}

		const slot = player.inventory.items.find(i => i.id === object.itemId);
		if (!slot || slot.quantity < object.quantity) {
			return false;
		}

		// Consume items on completion
		removeItem(player.inventory, object.itemId, object.quantity);
		return true;
	},

	destroy_npc(object, bus) {
		if (object.type !== 'destroy_npc') {
			return false;
		}

		for (const ev of bus.lastTickEvents) {
			if (ev.tag === EventTag.NpcDeath && 'npcType' in ev && ev.npcType === object.npcType) {
				object.killed++;
			}
		}

		return object.killed >= object.count;
	},

	wait_at(object, bus, player, _state, mapWidth, mapHeight) {
		if (object.type !== 'wait_at') {
			return false;
		}

		const d = torusDist(player.x, player.y, object.x, object.y, mapWidth, mapHeight);
		if (d > object.radius) {
			return false;
		}

		// Count hours from TimeAdvance events
		for (const ev of bus.lastTickEvents) {
			if (ev.tag === EventTag.TimeAdvance) {
				object.hoursWaited++;
			}
		}

		return object.hoursWaited >= object.hoursRequired;
	},

	interact_cell(object, bus) {
		if (object.type !== 'interact_cell') {
			return false;
		}

		for (const ev of bus.lastTickEvents) {
			if (ev.tag === EventTag.LandmarkChangeOwner
				&& 'landmarkId' in ev
				&& ev.landmarkId === object.x) {
				return true;
			}

			if (ev.tag === EventTag.WorldCellChange
				&& 'x' in ev && 'y' in ev
				&& ev.x === object.x && ev.y === object.y) {
				return true;
			}
		}

		return false;
	},
};

// ── Reward applier registry (data-driven) ──

type RewardApplier = (reward: QuestReward, player: PlayerState, bus: EventBus) => void;

const REWARD_APPLIERS: Record<string, RewardApplier> = {
	gold(reward, player, bus) {
		if (reward.type !== 'gold') {
			return;
		}

		player.gold += reward.amount;
		bus.emit({tag: EventTag.PlayerGoldChange, delta: reward.amount, newTotal: player.gold});
	},

	xp(reward, player) {
		if (reward.type !== 'xp') {
			return;
		}

		player.levelData.exp += reward.amount;
	},

	item(reward, player) {
		if (reward.type !== 'item') {
			return;
		}

		addItem(player.inventory, makeItem(reward.itemId, reward.quantity));
	},

	reputation(reward, player, bus) {
		if (reward.type !== 'reputation') {
			return;
		}

		const newValue = (player.reputation[reward.faction] ?? 0) + reward.delta;
		player.reputation[reward.faction] = newValue;
		bus.emit({
			tag: EventTag.ReputationChange,
			faction: reward.faction,
			delta: reward.delta,
			newValue,
		});
	},

	event(reward, _player, bus) {
		if (reward.type !== 'event') {
			return;
		}

		bus.emit(reward.event);
	},
};

// ── Quest engine ──

export type QuestTickContext = {
	bus: EventBus;
	player: PlayerState;
	state: GameState;
	mapWidth: number;
	mapHeight: number;
};

export class QuestEngine {
	/** Tick — check all active quest objectives, complete/fail as needed. */
	tick({bus, player, state, mapWidth, mapHeight}: QuestTickContext): void {
		const toComplete: number[] = [];

		for (let qi = player.activeQuests.length - 1; qi >= 0; qi--) {
			const quest = player.activeQuests[qi];

			// Check expiry
			if (quest.expireDay !== undefined && state.worldTime.day > quest.expireDay) {
				player.activeQuests.splice(qi, 1);
				player.completedQuestIds.push(quest.id); // Track as done (failed)
				bus.emit({tag: EventTag.QuestFail, questId: quest.id, reason: 'expired'});
				continue;
			}

			// Check each incomplete objective
			let allDone = true;
			let anyUpdated = false;
			for (const object of quest.objectives) {
				if (object.completed) {
					continue;
				}

				const checker = OBJECTIVE_CHECKERS[object.type];
				if (checker?.(object, bus, player, state, mapWidth, mapHeight)) {
					object.completed = true;
					anyUpdated = true;
				} else {
					allDone = false;
				}
			}

			if (anyUpdated && !allDone) {
				bus.emit({tag: EventTag.QuestUpdate, questId: quest.id});
			}

			if (allDone) {
				toComplete.push(qi);
			}
		}

		// Complete quests (reverse order to preserve indices)
		for (const qi of toComplete) {
			const quest = player.activeQuests[qi];
			player.activeQuests.splice(qi, 1);
			player.completedQuestIds.push(quest.id);
			// Apply rewards
			for (const reward of quest.rewards) {
				REWARD_APPLIERS[reward.type]?.(reward, player, bus);
			}

			bus.emit({tag: EventTag.QuestComplete, questId: quest.id});
		}
	}

	/** Accept a quest — move to active list + emit onAccept events. */
	accept(quest: Quest, player: PlayerState, bus: EventBus): void {
		player.activeQuests.push(quest);
		if (quest.onAccept) {
			for (const ev of quest.onAccept) {
				bus.emit(ev);
			}
		}

		bus.emit({tag: EventTag.QuestStart, questId: quest.id, title: quest.title});
	}

	/** Abandon an active quest. */
	abandon(questId: string, player: PlayerState): void {
		const idx = player.activeQuests.findIndex(q => q.id === questId);
		if (idx !== -1) {
			player.activeQuests.splice(idx, 1);
		}
	}

	/** Check if a quest is already active or completed. */
	isKnown(questId: string, player: PlayerState): boolean {
		return player.completedQuestIds.includes(questId)
			|| player.activeQuests.some(q => q.id === questId);
	}
}

// ── Helpers ──

function findSettlement(state: GameState, id: number): AnySettlement | undefined {
	return (state.settlements as AnySettlement[]).find(s => s.id === id)
		?? state.villages.find(v => v.id === id);
}

File: timaert_c/src/events/quests/quest_engine.cpp
#include "events/quests/quest_engine.h"
#include "core/torus.h"
#include <algorithm>
#include <cstddef>
#include <utility>

namespace sm {

namespace {

static bool obj_in_radius(const GameState& gs, float px, float py, float cx, float cy, float r) {
    return torus_dist_sq(px, py, cx, cy, float(gs.mapW), float(gs.mapH)) <= r * r;
}

static bool settlement_position(const GameState& gs, int id, float& x, float& y) {
    for (const auto& s : gs.settlements) {
        if (s.id == id) {
            x = float(s.x);
            y = float(s.y);
            return true;
        }
    }
    for (const auto& v : gs.villages) {
        if (v.id == id) {
            x = float(v.x);
            y = float(v.y);
            return true;
        }
    }
    return false;
}

static void push_unique_string(std::vector<std::string>& values, const std::string& value) {
    if (value.empty()) return;
    if (std::find(values.begin(), values.end(), value) == values.end()) {
        values.push_back(value);
    }
}

static void push_string(std::vector<std::string>& values, const std::string& value) {
    if (!value.empty()) values.push_back(value);
}

static void emit_reward(const Reward& r, PlayerState& p, EventBus& bus) {
    switch (r.kind) {
        case RewardKind::Gold: {
            p.gold += r.amount;
            GameEvent ev{EventTag::PlayerGoldChange};
            ev.ix = r.amount;
            ev.iy = p.gold;
            ev.b = kEventEffectAlreadyApplied;
            bus.emit(ev);
            break;
        }
        case RewardKind::Xp:
            p.sheet.levelData.exp += r.amount;
            break;
        case RewardKind::Item:
            p.inventory.add(r.itemId, r.amount);
            break;
        case RewardKind::Reputation: {
            p.reputation[r.faction] += r.delta;
            GameEvent ev{EventTag::ReputationChange};
            ev.s1 = r.faction;
            ev.ix = r.delta;
            ev.iy = p.reputation[r.faction];
            ev.b = kEventEffectAlreadyApplied;
            bus.emit(ev);
            break;
        }
        case RewardKind::Event:
            bus.emit(r.event);
            break;
    }
}

static bool eval_objective(Objective& o, const std::vector<GameEvent>& events,
                           GameState& gs) {
    if (o.completed) return true;
    PlayerState& p = gs.player;
    switch (o.kind) {
        case ObjectiveKind::VisitCell:
            if (obj_in_radius(gs, p.x, p.y, float(o.ix), float(o.iy), o.radius)) o.completed = true;
            break;
        case ObjectiveKind::FindLocation:
            for (auto& ev : events) {
                if (ev.tag == EventTag::PlayerMove &&
                    ev.ix == o.cellX && ev.iy == o.cellY) o.completed = true;
            }
            break;
        case ObjectiveKind::DeliverItems:
            {
                float sx = 0.0f, sy = 0.0f;
                if (settlement_position(gs, o.targetSettlementId, sx, sy)
                    && obj_in_radius(gs, p.x, p.y, sx, sy, 3.0f)
                    && p.inventory.count(o.itemId) >= o.quantity) {
                    o.completed = p.inventory.remove(o.itemId, o.quantity);
                }
            }
            break;
        case ObjectiveKind::DestroyNpc:
            for (auto& ev : events)
                if (ev.tag == EventTag::NpcDeath &&
                    (int(ev.a) == o.npcType || ev.ix == o.npcType)) o.killed++;
            if (o.killed >= o.count) o.completed = true;
            break;
        case ObjectiveKind::WaitAt:
            if (obj_in_radius(gs, p.x, p.y, float(o.ix), float(o.iy), o.radius)) {
                for (auto& ev : events) {
                    if (ev.tag == EventTag::TimeAdvance) {
                        ++o.hoursWaited;
                    }
                }
                if (o.hoursWaited >= o.hoursRequired) {
                    o.completed = true;
                }
            }
            break;
        case ObjectiveKind::InteractCell:
            for (auto& ev : events) {
                if (ev.tag == EventTag::WorldCellChange &&
                    ev.ix == o.ix && ev.iy == o.iy) o.completed = true;
                if (ev.tag == EventTag::LandmarkChangeOwner &&
                    (int(ev.a) == o.ix || ev.ix == o.ix)) o.completed = true;
            }
            break;
    }
    return o.completed;
}

} // namespace

void QuestEngine::tick(std::vector<Quest>& active, EventBus& bus, GameState& gs) {
    auto& events = bus.last_tick_events();
    std::vector<Quest> completed;
    completed.reserve(active.size());

    for (std::size_t qi = active.size(); qi > 0u; ) {
        --qi;
        Quest& q = active[qi];
        if (q.expireDay >= 0 && gs.worldTime.day > q.expireDay) {
            push_string(gs.player.completedQuestIds, q.id);
            push_unique_string(gs.player.failedQuestIds, q.id);
            GameEvent ev; ev.tag = EventTag::QuestFail; ev.s1 = q.id;
            ev.s2 = "expired";
            ev.a = std::uint32_t(quest_id_key(q.id));
            ev.b = kEventEffectAlreadyApplied;
            bus.emit(ev);
            active.erase(active.begin() + static_cast<std::ptrdiff_t>(qi));
            continue;
        }

        bool allDone = true;
        bool anyUpdated = false;
        for (auto& o : q.objectives) {
            const bool wasDone = o.completed;
            if (!eval_objective(o, events, gs)) allDone = false;
            if (!wasDone && o.completed) anyUpdated = true;
        }
        if (anyUpdated && !allDone) {
            GameEvent ev{EventTag::QuestUpdate};
            ev.s1 = q.id;
            ev.a = std::uint32_t(quest_id_key(q.id));
            bus.emit(ev);
        }
        if (allDone) {
            completed.push_back(std::move(q));
            active.erase(active.begin() + static_cast<std::ptrdiff_t>(qi));
        }
    }

    for (auto& q : completed) {
        push_string(gs.player.completedQuestIds, q.id);
        for (auto& r : q.rewards) emit_reward(r, gs.player, bus);
        GameEvent ev; ev.tag = EventTag::QuestComplete; ev.s1 = q.id;
        ev.a = std::uint32_t(quest_id_key(q.id));
        ev.b = kEventEffectAlreadyApplied;
        bus.emit(ev);
    }
}

void QuestEngine::accept(std::vector<Quest>& active,
                         Quest q,
                         const PlayerState& player,
                         EventBus& bus) {
    (void)player;
    active.push_back(std::move(q));
    const Quest& accepted = active.back();
    bus.emit_all(accepted.onAccept);
    GameEvent ev; ev.tag = EventTag::QuestStart; ev.s1 = accepted.id;
    ev.s2 = accepted.title;
    ev.a = std::uint32_t(quest_id_key(accepted.id));
    bus.emit(ev);
}

void QuestEngine::abandon(std::vector<Quest>& active, const std::string& id, EventBus& bus) {
    for (auto it = active.begin(); it != active.end(); ++it) {
        if (it->id == id) {
            GameEvent ev; ev.tag = EventTag::QuestFail; ev.s1 = id;
            ev.s2 = "abandoned";
            ev.a = std::uint32_t(quest_id_key(id));
            bus.emit(ev);
            active.erase(it);
            return;
        }
    }
}

bool QuestEngine::is_known(const std::vector<Quest>& active,
                           const PlayerState& player,
                           const std::string& id) const {
    for (const auto& q : active) {
        if (q.id == id) return true;
    }
    return std::find(player.completedQuestIds.begin(),
                     player.completedQuestIds.end(),
                     id) != player.completedQuestIds.end()
        || std::find(player.failedQuestIds.begin(),
                     player.failedQuestIds.end(),
                     id) != player.failedQuestIds.end();
}

} // namespace sm
